envoy-filters/mgw-source/filters/http/mgw/mgw.cc

```diff
@@ -0,0 +1,136 @@
+  set_body_ = readMetadata(&req_callbacks_->streamInfo().dynamicMetadata());
+  //if we do not modify the body we continue to other filters
+  if (!set_body_) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+  // if stream is not ended, stop continueing to other filters and buffer. 
+  if (!end_stream) {
+    return Http::FilterHeadersStatus::StopIteration;
+  }
+  //once the stream is complete (end_stream == true), we modify the payload.
+  setPayload(modified_body_, req_callbacks_->decodingBuffer(), req_headers_);
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_stream) {
+
+  // if we do not modify the body we continue to other filters
+  if (!set_body_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  ENVOY_LOG(trace, "decodeData with data = {} , end_stream = {}", data.toString(), end_stream);
+  // if stream is not ended, stop continueing to other filters and buffer.
+  if (!end_stream) {
+    return Http::FilterDataStatus::StopIterationAndBuffer;
+  }
+  // once the stream is complete (end_stream == true), we modify the payload.
+  setPayload(modified_body_, req_callbacks_->decodingBuffer(), req_headers_);
+  return Http::FilterDataStatus::Continue;
+}
+
+void Filter::setPayload(std::string new_payload, const Buffer::Instance* decoding_buffer,
+                     Http::RequestHeaderMap* req_headers) {
+  ENVOY_LOG(debug, "Modifying payload ...");
+  // create a new buffer instance with new payload data.
+  Buffer::OwnedImpl modified_body(new_payload);
+
+  //if we never got a payload from the downstream/ header only request.
+  if (decoding_buffer == nullptr) {
+    // We are not streaming back this this payload data so we set streaming_filter = false.
+    req_callbacks_->addDecodedData(modified_body, false);
+  }
+  else {
+    // Since we modifying the existing payload, we drain the data buffer and add new data.
+    req_callbacks_->modifyDecodingBuffer([&modified_body](Buffer::Instance& data) {
+      data.drain(data.length());
+      data.move(modified_body);
+    });
+  }
+  // Since we modified the payload, need to set new content length.
+  decoding_buffer = req_callbacks_->decodingBuffer();
+  req_headers->setContentLength(decoding_buffer->length());
+  ENVOY_LOG(debug, "payload successfully modified");
+}
+
+bool Filter::readMetadata(const envoy::config::core::v3::Metadata* metadata) {
+  std::string jsonJWTPayload;
+  // We read metadata under ext_authz filter and if find a key "payload", 
+  // then decide to modify the payload.
+  std::string PayloadMetadataKey = "payload";
+  // Reading metadata under external authorization
+  const auto* payload = &Config::Metadata::metadataValue(
+      metadata, HttpFilterNames::get().ExtAuthorization, PayloadMetadataKey);
+
+  if (payload != nullptr && (payload->kind_case() == ProtobufWkt::Value::kStringValue)) {
+    modified_body_ = payload->string_value();
```

Comment:
Storing the payload in a class level variable would cause issues in multithreaded approaches.

Reply:
Filter obj are initiated per request stream, this will not break multithreaded scenarios. referring https://www.envoyproxy.io/docs/envoy/latest/intro/life_of_a_request#http-filter-chain-processing